src/lib/ldap/filter.c

```diff
@@ -0,0 +1,565 @@
+		break;
+
+	/*
+	 *	Operator should not be unset at the end of a filter
+	 */
+	case LDAP_FILTER_OP_UNSET:
+		fr_assert(0);
+		break;
+	}
+
+	/*
+	 *	Take a copy of the original filter for debug output
+	 */
+	MEM(node->orig = talloc_zero_array(node, char, fr_sbuff_diff(sbuff, &marker) + 1));
+	memcpy(node->orig, fr_sbuff_current(&marker), fr_sbuff_diff(sbuff, &marker));
+	DEBUG3("%*sParsed LDAP filter (%s)", depth, "", node->orig);
+
+	return 0;
+}
+
+/** Parse individual LDAP filter nodes
+ *
+ * A node can either be a group of nodes joined with a logical operator
+ * or an individual filter.
+ *
+ * @param[in,out] node	to populate with parsed filter.
+ * @param[in] sbuff	pointing to filter to parse.
+ * @param[in] depth	to indent debug output, indicating nesting of groups.
+ * @param[in] attrs	listed in sync config.
+ * @return
+ *  - 0 on success
+ *  - < 0 on error
+ */
+static fr_slen_t ldap_filter_parse_node(ldap_filter_t *node, fr_sbuff_t *sbuff, int depth, char const **attrs)
+{
+	fr_sbuff_marker_t	marker;
+	fr_slen_t		ret;
+
+	static bool const	logical_op_chars[UINT8_MAX +1] = {
+					['!'] = true, ['&'] = true, ['|'] = true,
+				};
+
+	if (!fr_sbuff_next_if_char(sbuff, '(')) {
+		fr_strerror_const("Missing '('");
+		return fr_sbuff_error(sbuff);
+	}
+
+	/*
+	 *	Firstly, look for the characters which indicate the start of a group of filters
+	 *	to be combined with a logical operator.
+	 */
+	fr_sbuff_marker(&marker, sbuff);
+	if (fr_sbuff_adv_past_allowed(sbuff, 1, logical_op_chars, NULL)) {
+		fr_sbuff_set(sbuff, &marker);
+		ret = ldap_filter_parse_logic(node, sbuff, depth, attrs);
+	} else {
+		ret = ldap_filter_parse_filter(node, sbuff, depth, attrs);
+	}
+
+	if (ret < 0) return ret;
+
+	if (!fr_sbuff_next_if_char(sbuff, ')')) {
+		fr_strerror_const("Missing ')'");
+		return fr_sbuff_error(sbuff);
+	}
+
+	/*
+	 *	If we're at the very top level we should be at the end
+	 *	of the buffer
+	 */
+	if ((depth == 0) && (fr_sbuff_extend(sbuff))) {
+		fr_strerror_const("Extra data at the end of LDAP filter");
+		return fr_sbuff_error(sbuff);
+	}
+
+	return 0;
+}
+
+/** Parse an LDAP filter into its component nodes
+ *
+ * @param[in] ctx	to allocate nodes in.
+ * @param[out] root	of tree of parsed nodes.
+ * @param[in] filter	to parse.
+ * @param[in] attrs	specified in the sync config.  Used to warn user if the filter
+ *			references attributes not being requested.
+ * @return
+ *	- 0 on success
+ *	< 0 on failure
+ */
+int fr_ldap_filter_parse(TALLOC_CTX *ctx, ldap_filter_t **root, char const *filter, char const **attrs)
+{
+	fr_sbuff_t	sbuff;
+	fr_slen_t	ret;
+
+	fr_sbuff_init_in(&sbuff, filter, strlen(filter));
+
+	MEM(*root = talloc_zero(ctx, ldap_filter_t));
+
+	ret = ldap_filter_parse_node(*root, &sbuff, 0, attrs);
+	if (ret < 0) {
+		char	*spaces, *text;
+
+		fr_canonicalize_error(ctx, &spaces, &text, ret + 1, filter);
+		ERROR("Invalid LDAP filter: %s", text);
+		ERROR("                     %s^ - %s", spaces, fr_strerror());
```

Comment:
Use `EMARKER`

Reply:
Generally for parsing it ends up being more flexible to have the caller pass in the sbuff, and return an `fr_slen_t`, then the caller can use the appropriate logging function.  In this case it's probably better to use one of the cf_log_err functions so we get info about the config item containing the error.